ext/typeexpr/defaults_test.go

```diff
@@ -53,9 +53,9 @@ func TestDefaults_Apply(t *testing.T) {
 			value: cty.MapVal(map[string]cty.Value{
 				"a": cty.StringVal("foo"),
 			}),
-			want: cty.ObjectVal(map[string]cty.Value{
+			want: cty.MapVal(map[string]cty.Value{
```

Comment:
I feel a little unsure about this result, although perhaps this is an edge-case that doesn't matter in practice. :thinking: 

I think originally we thought that `Defaults.Type` was supposed to constrain the result of applying the defaults, so that the result would conform to this type even if the input doesn't. However, since `Defaults.Apply` can't report errors, it isn't really clear how it's supposed to deal with a mismatch like this.

In practice, our one current caller (Terraform) converts the given input to the target type constraint before calling `Defaults.Apply` anyway, and so we should never actually get into this situation where the type constraint calls for an object but the actual given value is a map: Terraform would already have converted the map to the appropriate object type or returned an error explaining why that isn't possible.

The comment on `Defaults.Apply` already says that the caller must do type conversion before calling it and so my sense is that this test case is effectively invalid: what it's passing in here violates the documented contract. Part of me thinks we should make `Defaults.Apply` panic if it can see that the contract isn't being upheld -- that is, if the given value doesn't already conform to the target type constraint -- but type conformance checks are relatively expensive and so we probably don't want to start sprinkling them in here willy-nilly.

We could perhaps add just a single type conformance check right up front in the public `Defaults.Apply`, before it delegates to the unexported `apply` method:

```go
func (d *Defaults) Apply(val cty.Value) cty.Value {
	if errs := val.Type().TestConformance(d.Type); len(errs) > 0 {
		panic("value does not conform to expected type")
	}
	return d.apply(val)
}
```

I don't think this is crucial but it would at least help give fast feedback if a caller isn't using this correctly, rather than having it just return something confusing.

(I'm proposing only to pre-check the _input_ here. The result might still not conform to the given type constraint if there are any non-unifyable collection elements in there, and that's up to the caller to handle somehow after we return, but I think it _is_ reasonable to expect that the caller at least does its homework of getting the input value into the prescribed shape beforehand.)

Reply:
I'm not going to make this change, instead I've reverted some previous documentation changes to say that users should go back to doing the conversion after instead of before.

We can/should come back and revisit this when we have an opportunity to change the API so that we can return errors instead of panicking.